ttnn/cpp/ttnn/types.hpp

```diff
@@ -59,76 +59,70 @@ struct CoreGrid {
     }
 };
 
-// Keep track of live buffers and the device addresses they were assigned.
-// When a buffer is created, it is provided a buffer_id using get_buf_id().
-// The address for this buffer is assigned to buffer_id when the buffer is asynchronously allocated.
-// When the buffer destructor is called, or the buffer is asynchronously deallocated, the worker thread
-// will look up the address for buffer_id to free memory on device.
-class buffer_address_map {
+// This buffer class is compatible with multithreaded runtime (which lives in tt_eager)
+// It is derived from the tt_metal::Buffer class, but defines its own asynchronous allocation functions
+class Buffer : public tt::tt_metal::Buffer, public std::enable_shared_from_this<Buffer> {
     public:
-        void insert(uint32_t buf_id, uint32_t buf_addr) {
-            std::scoped_lock<std::mutex> lock(this->map_mutex);
-            this->buf_id_to_address_map.insert({buf_id, buf_addr});
-        }
-        void erase(uint32_t buf_id) {
-            std::scoped_lock<std::mutex> lock(this->map_mutex);
-            this->buf_id_to_address_map.erase(buf_id);
-        }
-        uint32_t buffer_address(uint32_t buf_id) {
-            std::scoped_lock<std::mutex> lock(this->map_mutex);
-            return this->buf_id_to_address_map.at(buf_id);
+        static std::shared_ptr<Buffer> Create(Device* device, uint64_t size, uint64_t page_size, const BufferType buffer_type,
+                const TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED,
+                std::optional<ShardSpecBuffer> shard_parameters = std::nullopt
+            ) {
+                // Factory Method for creating Buffer objects. Will return std::shared_ptr<Buffer>.
+                // This is the only interface for creating Buffers, to ensure that users don't use
+                // the Buffer class without shared_ptr, since this class uses enable_shared_from_this.
+                std::shared_ptr<Buffer> buffer_instance(new Buffer(device, size, page_size, buffer_type, buffer_layout, shard_parameters));
+                buffer_instance->allocate(); // Asynchronously allocate the buffer
+                return buffer_instance;
+            }
+        ~Buffer() {
+            this->deallocate(); // Asynchronously clean up device state used by this buffer
         }
-        uint32_t get_buf_id() {
-            return buf_id++;
+        uint32_t address() {
+            while (not this->allocated_); // Wait until buffer is allocated
```

Comment:
Please never write a loop like this.
Please use simple this->allocated_.wait(true);

Reply:
Done.